core/state/statedb.go

```diff
@@ -1071,7 +1072,7 @@ func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
 	}
 	if root != origin {
 		start := time.Now()
-		if err := s.db.TrieDB().Update(nodes); err != nil {
+		if err := s.db.TrieDB().Update(root, origin, nodes); err != nil {
```

Comment:
Here origin will be `types.EmptyRootHash` for an empty tree but much of your diff uses `common.Hash{}` to signal the same thing. Technically `types.EmptyRootHash` seems the correct one and whilst we could perhaps have the empty hash mean the same thing, I'm unsure we should permit that leniency. Maybe I'd rather enforce the proper empty root hash everywhere?

Reply:
Yep